datadog_checks_base/datadog_checks/base/utils/db/statement_metrics.py

```diff
@@ -0,0 +1,127 @@
+# (C) Datadog, Inc. 2020-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class StatementMetrics:
+    """
+    This class supports normalized statement-level metrics, which are collected from the database's
+    statistics tables, ex:
+
+        - Postgres: pg_stat_statements
+        - MySQL: performance_schema.events_statements_summary_by_digest
+        - Oracle: V$SQLAREA
+        - SQL Server: sys.dm_exec_query_stats
+        - DB2: mon_db_summary
+
+    These tables are monotonically increasing.
+    """
+
+    def __init__(self):
+        self.previous_statements = {}
+
+    def compute_derivative_rows(self, rows, metrics, key):
+        """
+        Compute the first derivative of column-based metrics for a given set of rows. This also resets the
+        statement cache so should only be called once per check run.
+
+        - **rows** (_List[dict]_) - rows from current check run
+        - **metrics** (_List[str]_) - the metrics to compute for each row
+        - **key** (_callable_) - function for an ID which uniquely identifies a row across runs
+        """
+        result = []
+        new_cache = {}
+        negative_result_found = False
+        metrics = set(metrics)
+        if len(rows) > 0:
+            dropped_metrics = metrics - set(rows[0].keys())
+            if dropped_metrics:
+                logger.warning(
+                    'Some statement metrics are not available from the table: %s', ','.join(m for m in dropped_metrics)
+                )
+
+        for row in rows:
+            row_key = key(row)
+            if row_key in new_cache:
+                logger.debug(
+                    'Collision in cached query metrics. Dropping existing row, row_key=%s new=%s dropped=%s',
+                    row_key,
+                    row,
+                    new_cache[row_key],
+                )
+            new_cache[row_key] = row
+            prev = self.previous_statements.get(row_key)
+            if prev is None:
+                continue
+            metric_columns = metrics & set(row.keys())
+            if any([row[k] - prev[k] < 0 for k in metric_columns]):
+                # The table was truncated or stats reset; begin tracking again from this point
+                negative_result_found = True
+                continue
+            if all([row[k] - prev[k] == 0 for k in metric_columns]):
+                # No metrics to report; query did not run
+                continue
+            derived = {k: row[k] - prev[k] if k in metric_columns else row[k] for k in row.keys()}
```

Comment:
(Nit) A few nits:

- Add empty lines between groups of related statements to ease readability.
- Compute `row[k] - prev[k]` differences once and for all, then use that piece of info to decide what to do (motivation is not performance, but readability: easier to reason with a variable called `metric_values` when what we're doing is effectively make a decision based on whether there's any valid/interesting metric to report)
- Add a few more motivational comments for what I initially considered to be unclear (i.e. that the values we're submitting is effectively just the difference between the current and previous state, and that the "derived" dict is the combination of these "diff" metrics, as well as values of unknown columns in the row).

```suggestion
            metric_columns = metrics & set(row.keys())

            # By default we'll submit differences between the current 
            # state and the previous state.
            metric_values = {k: row[k] - prev[k] for k in metric_columns}

            if any(value < 0 for value in metric_values.values()):
                # The table was truncated or stats reset; begin tracking again from this point
                negative_result_found = True
                continue

            if all(value == 0 for value in metric_values.values()):
                # No metrics to report; query did not run
                continue

            # NOTE: values of unknown columns are submitted as-is.
            derived = {k: metric_value if k in metric_values else v for k, v in row.items()}
```

Reply:
I cleaned up the code a bit and I think it addresses all of your concerns, but let me know if that's not the case! I added a lot more documentation to explain both what the code is doing and the reason for the very particular requirements, perhaps at the risk of going overboard on comment length. But given the number of people confused by it, it might be justified. Let me know!